bittide-instances/src/Project/Handle.hs

```diff
@@ -37,15 +40,45 @@ expectLine = expectLine' ""
         Stop (Error msg) -> do
           putStrLn s1
           assertFailure msg
+expectLine (Just (timeout, msg)) h f = do
+  initTime <- getCurrentTime
+  expectLine' initTime "" h f
+ where
+  timeout' = toRational timeout
+  expectLine' t0 s0 h' f' = do
+    now <- getCurrentTime
+    if toRational (diffUTCTime t0 now) >= timeout'
+      then errorIO $ "expect line timed out on: `" <> msg <> "`. previously read: \n" <> s0
+      else do
+        line <- hGetLine h
+        let
+          trimmed = trim line
+          s1 = s0 <> "\n" <> line
```

Comment:
I think I'd rather see a list of lines  of type `[String]` rather than reading on a per line basis and then manually re-inserting new lines. You can easily convert `[String]` to `String` using `unlines`.